### PR TITLE
[AMDGPU][GFX12] Add 16 bit atomic fadd instructions

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**
- image_atomic_pk_add_f16
- image_atomic_pk_add_bf16
- ds_pk_add_bf16
- ds_pk_add_f16
- ds_pk_add_rtn_bf16
- ds_pk_add_rtn_f16
- flat_atomic_pk_add_f16
- flat_atomic_pk_add_bf16
- global_atomic_pk_add_f16
- global_atomic_pk_add_bf16
- buffer_atomic_pk_add_f16
- buffer_atomic_pk_add_bf16

### CODE REVIEW

File: clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// RUN: %clang_cc1 -O0 -cl-std=CL2.0 -triple amdgcn-amd-amdhsa -target-cpu gfx1200 \
+// RUN:   %s -S -emit-llvm -o - | FileCheck %s
+
+// RUN: %clang_cc1 -O0 -cl-std=CL2.0 -triple amdgcn-amd-amdhsa -target-cpu gfx1200 \
```
Comment (arsenm) [inline clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl:4]: Don't see the point in running this all the way to codegen

File: clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// RUN:   -S -o - %s | FileCheck -check-prefix=GFX12 %s
+
```
Comment (arsenm) [inline clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl:6]: Missing the test that shows these are rejected on other targets?

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 3]
```diff
@@ -1368,6 +1391,28 @@ def int_amdgcn_struct_ptr_buffer_atomic_cmpswap : Intrinsic<
 // gfx908 intrinsic
 def int_amdgcn_struct_buffer_atomic_fadd : AMDGPUStructBufferAtomic<llvm_anyfloat_ty>;
 def int_amdgcn_struct_ptr_buffer_atomic_fadd : AMDGPUStructPtrBufferAtomic<llvm_anyfloat_ty>;
+// gfx12 intrinsic
+def int_amdgcn_struct_buffer_atomic_fadd_v2bf16 : Intrinsic <
+  [llvm_v2i16_ty],
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1396]: Probably should start using v2bf16 for these, although we have the bad precedent of not using it for the existing operations (I'm currently working on fixing this for some other intrinsics)

**Review by arsenm — COMMENTED** [review]
Seems to be missing atomicrmw fadd support?

File: clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl (hunk 1 above)
Comment (mariusz-sikora-at-amd) [reply clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl:4]: What do you suggest ? I copied this test from other builtins-fp-atomics-gfxXX.
I thought this is a good test which covers both llvm intrinsic and ISA generation.

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 3 above)
Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1396]: Patch to v2bf16 legal is committed, you can now update this to use it

File: clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl (hunk 2 above)
Comment (mariusz-sikora-at-amd) [reply clang/test/CodeGenOpenCL/builtins-fp-atomics-gfx12.cl:6]: Added here: https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenOpenCL/builtins-amdgcn-fp-atomics-gfx11-err.cl

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll [hunk 4]
```diff
@@ -27,34 +27,23 @@ main_body:
...
-}
-
-define amdgpu_ps float @atomic_pk_add_bf16_1d_v4(<8 x i32> inreg %rsrc, <4 x i16> %data, i32 %s) {
-; GFX12-LABEL: atomic_pk_add_bf16_1d_v4:
-; GFX12:       ; %bb.0: ; %main_body
-; GFX12-NEXT:    image_atomic_pk_add_bf16 v[0:1], v2, s[0:7] dmask:0x3 dim:SQ_RSRC_IMG_1D th:TH_ATOMIC_RETURN
-; GFX12-NEXT:    s_waitcnt vmcnt(0)
-; GFX12-NEXT:    ; return to shader part epilog
-main_body:
-  %out = call <4 x i16> @llvm.amdgcn.image.atomic.pk.add.bf16.1d.v4i16.v4i16(<4 x i16> %data, i32 %s, <8 x i32> %rsrc, i32 0, i32 0)
```
Comment (mariusz-sikora-at-amd) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll]: Found issue with <4 x bfloat> and GlobalISel. I will try to debug this and prepare fix.

**Comment by mariusz-sikora-at-amd** [comment]
What is the plan for atomic_{flat/ds/global}_bf16 builtins ? Right now they are accepting <2 x i16> instead of <2 x bfloat>. Do we want to create new builtins or we want to override them to accept both <2 x i16> and <2 x bfloat> ?

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll (hunk 4 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll]: Is it #77448?

Comment (mariusz-sikora-at-amd) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll]: Yes, #77448 will fix these issue.

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 3 above)
Comment (mariusz-sikora-at-amd) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td]: Done

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll (hunk 4 above)
Comment (mariusz-sikora-at-amd) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll]: Added once again 4xbfloat tests

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll [hunk 5]
```diff
@@ -44,6 +44,14 @@ main_body:
   ret float 1.0
 }
 
+define amdgpu_ps float @atomic_pk_add_bf16_1d_v4(<8 x i32> inreg %rsrc, <4 x bfloat> %data, i32 %s) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll:47]: Should also include a test with no use of the return value, plus with some cache bits set

**Comment by arsenm** [comment]
> What is the plan for atomic_{flat/ds/global}_bf16 builtins ? Right now they are accepting <2 x i16> instead of <2 x bfloat>. Do we want to create new builtins or we want to override them to accept both <2 x i16> and <2 x bfloat> ?

I've been planning on looking into it. Currently leaning towards creating new builtins

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll [hunk 6]
```diff
@@ -1,56 +1,244 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
-; RUN: llc -march=amdgcn -mcpu=gfx1200 -verify-machineinstrs < %s | FileCheck -check-prefix=GFX12 %s
-; RUN: llc -march=amdgcn -global-isel=1 -mcpu=gfx1200 -verify-machineinstrs < %s | FileCheck -check-prefix=GFX12 %s
+; RUN: llc -march=amdgcn -mcpu=gfx1200 -verify-machineinstrs < %s | FileCheck -check-prefix=GFX12-SDAG %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll:2]: Use an explicit -global-isel=0 for the dag line

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll (hunk 6 above)
Comment (mariusz-sikora-at-amd) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.image.atomic.pk.add.ll]: Done

**Pull request merged**